### ISSUE TITLE
question about the code 

### ISSUE THREAD

**Issue by AlanYeg** [abertsch72/unlimiformer#14]
Hi,dear developers
the query variable in attention_forward_hook, the comment says it has shape # query: (batch, beam, head, dim)
```

def attention_forward_hook(self, module, input, output):
        # output: (batch, time, 3 * heads * attention_dim)
        if self.is_input_encoding_pass or self.is_first_test_decoding_step:
            return
        with torch.no_grad():
            query = self.process_query(output)[:,-1] # (batch * beam, head, dim)
            query = query[:, self.head_nums] # (batch * beam, head, dim)
            if self.normalize:
                query = torch.nn.functional.normalize(query, dim=-1)

            if self.use_datastore:
                # query: (batch, beam, head, dim)

```
what does the beam index mean?

**Comment by urialon** [comment]
Hi @AlanYeg ,
Thank you for your interest in our work!

This code runs at test time, and the `beam` dimension is the number of beam-search beams.
This ensures that if you use beam search, every decoded beam is using Unlimiformer independently (that is, every decoded beam selects its own keys to retrieve).

I hope it helps. Does it make sense?
Best,
Uri

**Comment by urialon** [comment]
I'm closing this issue just to make sure we respond to all issues, but feel free to re-open it if you have any questions!
